Source/Parallelization/WarpXComm.cpp

```diff
@@ -1283,13 +1284,15 @@ void WarpX::NodalSyncPML (int lev, PatchType patch_type)
 {
     if (pml[lev]->ok())
     {
-        const auto& pml_E = (patch_type == PatchType::fine) ? pml[lev]->GetE_fp() : pml[lev]->GetE_cp();
-        const auto& pml_B = (patch_type == PatchType::fine) ? pml[lev]->GetB_fp() : pml[lev]->GetB_cp();
-        const auto& pml_F = (patch_type == PatchType::fine) ? pml[lev]->GetF_fp() : pml[lev]->GetF_cp();
-        const auto& pml_G = (patch_type == PatchType::fine) ? pml[lev]->GetG_fp() : pml[lev]->GetG_cp();
+        const std::array<amrex::MultiFab*,3>& pml_E = (patch_type == PatchType::fine) ?
+                                                      pml[lev]->GetE_fp() : pml[lev]->GetE_cp();
+        const std::array<amrex::MultiFab*,3>& pml_B = (patch_type == PatchType::fine) ?
+                                                      pml[lev]->GetB_fp() : pml[lev]->GetB_cp();
+        amrex::MultiFab* const& pml_F = (patch_type == PatchType::fine) ? pml[lev]->GetF_fp() : pml[lev]->GetF_cp();
+        amrex::MultiFab* const& pml_G = (patch_type == PatchType::fine) ? pml[lev]->GetG_fp() : pml[lev]->GetG_cp();
```

Comment:
what is the difference between adding `const`to the extreme left versus after the var-type

Reply:
That's a good question. 

In general, I think `const T &` and `T const &` are equivalent. The difficulty comes when we mix pointers with it, I think.

The reason why I moved `const` from left to right in the last two declarations is that the compiler did not like any other variant I tried. While it did not complain about the similar declarations above that use `std::array` on top of the `amrex::MultiFab*` pointer (aka, I could leave `const` on the left).

Frankly, I did not know well how to handle this. But what we have now maybe isn't the best way to go. No idea what `auto` was doing under the hood before, given that these PML getters do return `amrex::MultiFab*` pointers.

Maybe @WeiqunZhang and @ax3l can chime in.

The good news is that by making these kinds of changes where the types are explicit we might learn something as a side effect indeed.

Reply:
It's easier to understand if one reads it from right to left.  For example `MultiFab* const&` means a reference to a constant variable that is a pointer to MultiFab.  `const MultiFab* &` and `MultiFab const* &` mean a reference to a variable that is a pointer to a MultiFab that is constant.  The latter does not work in this case because `const *` means the MultiFab object is a constant, whereas `* const` mean the pointer variable is a constant but MultiFab is not.  In this particular case, it's probably cleaner to do `amrex::MultiFab* const = ...` without the reference because it's a simple copy of a pointer, not a large object.

Reply:
Thank you for your help, Weiqun! I committed your suggestion in https://github.com/ECP-WarpX/WarpX/pull/2578/commits/fb8d60adc6d0da02b32466063d45dc0ed960a75d.

Reply:
All the above is correct.
If it was me, "easy-side const" all the way, due to the reasons Weiqun mentioned (more consistent to read).